File: app/src/main/java/com/ntrade/demo/view/bessel_curve/MScrollView.java
package com.ntrade.demo.view.bessel_curve;

import android.content.Context;
import android.util.AttributeSet;
import android.view.MotionEvent;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.widget.NestedScrollView;


/**
 * 创建者：leiwu
 * * 时间：2022/8/26 08:50
 * * 类描述：不拦截横向滑动的ScrollView
 * * 修改人：
 * * 修改时间：
 * * 修改备注：
 */
public class MScrollView extends NestedScrollView {

    private boolean isOldEvent;

    public MScrollView(@NonNull Context context) {
        super(context);
    }

    public MScrollView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public MScrollView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    private float downX;
    private float downY;

    @Override
    public boolean onInterceptTouchEvent(MotionEvent ev) {
        if (isOldEvent) {
            return false;
        }
        switch (ev.getAction()) {
            case MotionEvent.ACTION_DOWN:
                downX = ev.getX();
                downY = ev.getY();
                break;
            case MotionEvent.ACTION_MOVE:
                float moveX = ev.getX();
                float moveY = ev.getY();
                float difX = Math.abs(moveX - downX);
                float difY = Math.abs(moveY - downY);
                if (difX > difY) {
                    downX = moveX;
                    downY = moveY;
                    isOldEvent = true;
                    return false;
                }
                downX = moveX;
                downY = moveY;
                break;
            case MotionEvent.ACTION_UP:
                break;
        }
        return super.onInterceptTouchEvent(ev);
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent ev) {
        switch (ev.getAction()) {
            case MotionEvent.ACTION_UP:
                isOldEvent = false;
                break;
        }

        return super.dispatchTouchEvent(ev);
    }


}

File: app/src/main/java/com/ntrade/demo/view/scroll_view/MScrollView.kt
package com.ntrade.demo.view.scroll_view

import android.content.Context
import android.util.AttributeSet
import android.util.Log
import android.view.MotionEvent
import android.widget.ScrollView

/** * 创建者：wanglei
 * <p>时间：2022/1/27 11:33
 * <p>类描述：
 * <p>修改人：
 * <p>修改时间：
 * <p>修改备注：
 */
class MScrollView : ScrollView {
    constructor(context: Context?) : super(context)

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
    }

    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        mWidth = w;
        mHeight = h;
    }

    private var mWidth = 0//控件的宽
    private var mHeight = 0//控件的高
    private var downX = 0f
    private var downY = 0f
    private val flag by lazy { mWidth * 0.04f }
    private var isIntercept: Boolean = true
    private var isMove = false
    override fun dispatchTouchEvent(event: MotionEvent?): Boolean {
        event?.apply {
            when (action) {
                MotionEvent.ACTION_DOWN -> {
                    downX = x
                    downY = y
                    isIntercept = true
                }
                MotionEvent.ACTION_MOVE -> {
                    val moveX = Math.abs(x - downX)
                    val moveY = Math.abs(y - downY)

                    isIntercept =
                        when {
                            isMove -> isIntercept
                            moveY > flag && moveX < flag -> {
                                isMove = true
                                true
                            }
                            moveX > flag && moveY < flag -> {
                                isMove = true
                                false
                            }
                            moveY > flag && moveX > flag -> {
                                isMove = true
                                moveY > moveX
                            }
                            else -> true
                        }
                }
                MotionEvent.ACTION_UP -> isMove = false
            }
        }
        Log.d("MViewPage", "isIntercept = $isIntercept")
        return if (isIntercept) super.dispatchTouchEvent(event)
        else isIntercept
    }

    override fun onInterceptTouchEvent(ev: MotionEvent?): Boolean {
        Log.d("MViewPage", "onInterceptTouchEvent   isIntercept = $isIntercept")
        return isIntercept
    }

    private fun log(str: String) {
        Log.d("MViewPage", str)
    }
}